[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61921
* #61691
* #61589
* __->__ #61570

Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

supriyar committed:
[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)
Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

soulitzer commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5296,6 +5296,15 @@
 - func: _fake_quantize_learnable_per_channel_affine_backward(Tensor grad, Tensor self, Tensor scale, Tensor zero_point, int axis, int quant_min, int quant_max, float grad_factor=1.0) -> (Tensor, Tensor, Tensor)
   variants: function
 
+- func: fused_moving_avg_obs_fake_quant(Tensor(a) self, Tensor observer_on, Tensor fake_quant_on, Tensor averaging_const, Tensor(a!) running_min, Tensor(b!) running_max, Tensor(c!) scale, Tensor(d!) zero_point, int quant_min, int quant_max, int ch_axis, bool per_row_fake_quant=False, bool symmetric_quant=False) -> Tensor
```
To indicate that `self` is aliased with the output you could write: `fn_name(Tensor(a) self) -> Tensor(a)`.
You might want to shift `Tensor(a!) running_min, Tensor(b!) running_max, Tensor(c!) scale, Tensor(d!)` these letters down to `b, c, d, e`.

See https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native

soulitzer commented on aten/src/ATen/native/quantized/cpu/fused_obs_fake_quant.cpp:
```diff
@@ -0,0 +1,206 @@
+  }
+  // Calculate qparams and fake_quantize
+  auto fake_quant = fake_quant_on.item().toInt();
+  if (fake_quant) {
```
You probably want to factor your helper function so that it never returns a view by moving the `if fake_quant` logic in to the composite op instead (`fused_moving_avg_obs_fake_quant`).

The reason why you needed the don't enforce skip in `gen_variable_type` is because you return an alias here but did not declare the it as a view in `gen_inplace_or_view`. However, there are also situations e.g., when `fake_quant` is false, where you return a copy, so declaring this function as a view isn't entirely correct either.

So the rule in general is that a non-composite op should either consistently return a view or non-view. Factoring the original op into a composite wrapper + helper in this case allows us to shift that view/non-view inconsistency into the composite op, which skips the autograd kernel and thus forgoes the check for storage and tensor use count.

supriyar replied:
Since this involved creating multiple helper functions and break down of composite vs non-composite ops, I'm returning `x.clone()` instead when `fake_quant` is off. Running some jobs now to see if there is any impact due to this.

soulitzer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -225,7 +225,7 @@
     '_embedding_bag', '_embedding_bag_forward_only',
     'q_per_channel_scales', 'q_per_channel_zero_points',
-    'lu_unpack', '_cudnn_rnn_backward',
+    'lu_unpack', '_cudnn_rnn_backward', '_fused_moving_avg_obs_fq_helper',
```
We can probably remove this

soulitzer reviewed (commented):
Should be good once you move the `if` logic to the composite (wrapper) op. See my comment for details.

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5296,6 +5296,15 @@
+- func: _fused_moving_avg_obs_fq_helper(Tensor self, Tensor observer_on, Tensor fake_quant_on, Tensor averaging_const, Tensor(a!) running_min, Tensor(b!) running_max, Tensor(c!) scale, Tensor(d!) zero_point, int quant_min, int quant_max, int ch_axis, bool per_row_fake_quant=False, bool symmetric_quant=False) -> (Tensor output, Tensor mask)
+  dispatch:
+    CPU: fused_moving_avg_obs_fake_quant_cpu
+    #CUDA: fused_moving_avg_obs_fake_quant_cuda
```
This should be removed?

supriyar replied:
I added support for this in the next PR #61589

albanD replied:
Ho ok, I guess you can keep it then.

albanD commented:
The autograd side looks good to me :)

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

vkuzo commented on aten/src/ATen/native/quantized/cpu/fused_obs_fake_quant.cpp:
```diff
@@ -0,0 +1,206 @@
+          x_max_data[i],
+          qmin,
+          qmax,
+          symmetric_quant,
```
nit: comment on same line?

vkuzo commented on aten/src/ATen/native/quantized/cpu/fused_obs_fake_quant.cpp:
```diff
@@ -0,0 +1,206 @@
+          symmetric_quant,
+          /* preserve sparsity */ // Set preserve_sparsity to true for
+                                  // symmetric quantization.
+          false /* force power of two*/);
```
optional nit: `//` for comments, to allow people to use `/* ... */` to comment out chunks of code?

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a new fused MovingAvg Obs + FakeQuant operator(CPU)"


Summary:
Fused operator that computes moving average min/max values (in-place) of the input tensor and fake-quantizes it.
It expects the qmin/qmax values to reflect the range of the quantized tensor (instead of reduce_range)

Motivation for adding this operator is for performance reasons, since moving the computation from python to C++/CUDA can increase the performance of QAT.

Test Plan:
python test/test_quantization.py TestFusedObsFakeQuant

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29682762](https://our.internmc.facebook.com/intern/diff/D29682762)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29682762).
